### PR TITLE
Update integer decoding tests to use a symmetrical version of the encoder

### PR DESCRIPTION
**Pull request by HarelM**
This adds some encoding methods to allow symmetrical testing.
I have not written encoder to all decoders, but almost to all of them.
There's a lot of AI involved in writing the encoders I must admit, but I think the tests are helping to make sure both encoder and decoders are working as expected.

### CODE REVIEW

**Comment by DoFabien** [comment]
Hi! I've been testing the symmetric encoder infrastructure from this PR by implementing round-trip tests for the **Present Stream**.

**What I validated:**
-  Round-trip encode/decode with mixed presence masks
-  Edge cases (all-true, all-false sequences)
-  Large sequences (1000+ values)
-  **Offset tracking when concatenating multiple streams** — this catches subtle `byteLength` bugs that isolated tests miss

All tests pass.

<details>
<summary>📄 Click to expand the test snippet</summary>

```typescript
import { describe, expect, it } from "vitest";

import { PhysicalLevelTechnique } from "../metadata/tile/physicalLevelTechnique";
import { PhysicalStreamType } from "../metadata/tile/physicalStreamType";
import { decodeStreamMetadata } from "../metadata/tile/streamMetadataDecoder";
import BitVector from "../vector/flat/bitVector";
import { encodeBooleanRle } from "../encoding/encodingUtils";
import { decodeBooleanRle } from "./decodingUtils";
import { concatenateBuffers, createStream } from "./decodingTestUtils";
import IntWrapper from "./intWrapper";

function encodePresentStream(values: boolean[]): Uint8Array {
    return createStream(PhysicalStreamType.PRESENT, encodeBooleanRle(values), {
        technique: PhysicalLevelTechnique.VARINT,
        count: values.length,
    });
}

function roundTripPresentStream(values: boolean[]): { decodedValues: boolean[]; offset: number; streamLength: number } {
    const stream = encodePresentStream(values);
    const offset = new IntWrapper(0);
    const metadata = decodeStreamMetadata(stream, offset);

    const streamDataStart = offset.get();
    const decodedBytes = decodeBooleanRle(stream, metadata.numValues, offset);

    // Advance to the end of the stream using metadata.byteLength, matching prod decoders.
    offset.set(streamDataStart + metadata.byteLength);

    const bitVector = new BitVector(decodedBytes, metadata.numValues);
    const decodedValues = Array.from({ length: metadata.numValues }, (_, i) => bitVector.get(i));

    return { decodedValues, offset: offset.get(), streamLength: stream.length };
}

describe("present stream round-trip", () => {
    it("round-trips a mixed presence mask", () => {
        const values = [true, false, true, true, false, false, true, false, true, false, true];

        const { decodedValues, offset, streamLength } = roundTripPresentStream(values);

        expect(decodedValues).toEqual(values);
        expect(offset).toBe(streamLength);
    });

    it("handles all-true and all-false sequences", () => {
        const count = 100;
        const cases = [new Array(count).fill(true), new Array(count).fill(false)];

        for (const values of cases) {
            const { decodedValues, offset, streamLength } = roundTripPresentStream(values);

            expect(decodedValues).toEqual(values);
            expect(offset).toBe(streamLength);
        }
    });

    it("handles large sequences", () => {
        const values = new Array(1000).fill(true);

        const { decodedValues, offset, streamLength } = roundTripPresentStream(values);

        expect(decodedValues).toEqual(values);
        expect(offset).toBe(streamLength);
    });

    it("correctly advances offset when followed by another stream", () => {
        const presentValues = [true, false, true, false];
        const dataBytes = new Uint8Array([42, 100, 255, 0]);

        // Create both streams
        const presentStream = encodePresentStream(presentValues);
        const dataStream = createStream(PhysicalStreamType.DATA, dataBytes, {
            technique: PhysicalLevelTechnique.NONE,
            count: dataBytes.length,
        });

        // Concatenate: present stream followed by data stream
        const combined = concatenateBuffers(presentStream, dataStream);
        const offset = new IntWrapper(0);

        // Decode first stream (present)
        const meta1 = decodeStreamMetadata(combined, offset);
        const start1 = offset.get();
        decodeBooleanRle(combined, meta1.numValues, offset);
        offset.set(start1 + meta1.byteLength);

        // Decode second stream (data)
        const meta2 = decodeStreamMetadata(combined, offset);
        const start2 = offset.get();
        const decodedData = combined.slice(start2, start2 + meta2.byteLength);
        offset.set(start2 + meta2.byteLength);

        // Verify offset consumed entire buffer
        expect(offset.get()).toBe(combined.length);
        // Verify data stream content is intact
        expect(Array.from(decodedData)).toEqual(Array.from(dataBytes));
    });
});
```

</details>

I can open a follow-up PR with this test file once this PR is merged, or amend it here if you prefer. Let me know!

**Comment by HarelM** [comment]
I had a problem with the nullability stuff which I asked @Turtelll to take a look. He opened a PR to help me with it. I hope to continue the effort on that front next week.

You can open a PR against my branch if you want, or wait for this to be merged and then open a PR, your call.

File: ts/src/encoding/packNullableUtils.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import { type TypedArrayConstructor, type TypedArrayInstance } from "../decoding/unpackNullableUtils";
+import BitVector from "../vector/flat/bitVector";
+
+export function packNullable<T extends TypedArrayInstance>(data: T, presentBits: BitVector | null): T {
```
Comment (louwers) [inline ts/src/encoding/packNullableUtils.ts:4]: Missing test coverage.

File: ts/src/encoding/packNullableUtils.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    for (let i = 0; i < size; i++) {
+        if (presentBits.get(i)) {
+            result[counter++] = data[i];
+        }
+    }
+
+    return result;
+}
+
+export function packNullableBoolean(data: Uint8Array, dataSize: number, presentBits: BitVector | null): Uint8Array {
```
Comment (louwers) [inline ts/src/encoding/packNullableUtils.ts:36]: Missing test coverage.

File: ts/src/encoding/integerEncodingUtils.ts [hunk 3]
```diff
@@ -0,0 +1,792 @@
+import BitVector from "../vector/flat/bitVector";
+import IntWrapper from "../decoding/intWrapper";
+
+export function encodeVarintInt32Value(value: number, dst: Uint8Array, offset: IntWrapper): void {
```
Comment (louwers) [inline ts/src/encoding/integerEncodingUtils.ts:4]: Honestly, I feel like these should exist in a separate library.

It is not really MLT specific is it?

File: ts/src/encoding/packNullableUtils.ts (hunk 1 above)
Comment (louwers) [reply ts/src/encoding/packNullableUtils.ts:4]: Edit: nevermind.

File: ts/src/decoding/unpackNullableUtils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+describe("nullableUtils - non-nullable case", () => {
+    describe("unpackNullable", () => {
+        it("should return original array when presentBits is null", () => {
+            const dataStream = new Int32Array([1, 2, 3]);
+            const packed = packNullable(dataStream, null);
+            const result = unpackNullable(packed, null, 0);
+
+            expect(result).toBe(dataStream);
+            expect(result).toEqual(new Int32Array([1, 2, 3]));
```
Comment (louwers) [inline ts/src/decoding/unpackNullableUtils.spec.ts:14]: All these `toEqual` expectations seem redundant?

File: ts/src/decoding/unpackNullableUtils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
+import { describe, it, expect } from "vitest";
+import { unpackNullable, unpackNullableBoolean, unpackWithRepeat } from "./unpackNullableUtils";
+import BitVector from "../vector/flat/bitVector";
+import { packNullable, packNullableBoolean } from "../encoding/packNullableUtils";
+
+describe("nullableUtils - non-nullable case", () => {
+    describe("unpackNullable", () => {
+        it("should return original array when presentBits is null", () => {
```
Comment (louwers) [inline ts/src/decoding/unpackNullableUtils.spec.ts]: Nit: I think the descriptions should be "**the** original array" (etc.) to be grammatical.

File: ts/src/decoding/integerStreamDecoder.ts [hunk 6]
```diff
@@ -154,38 +153,72 @@ export function decodeConstLongStream(
...
-        return isSigned ? decodeZigZagValueInt64(value) : value;
+        return isSigned ? decodeZigZagInt64Value(value) : value;
     }
 
     return isSigned ? decodeZigZagConstRleInt64(values) : decodeUnsignedConstRleInt64(values);
 }
 
-function decodeIntBuffer(
+/*
+ * Currently the encoder uses only fixed combinations of encodings.
```
Comment (louwers) [inline ts/src/decoding/integerStreamDecoder.ts:164]: This should not be a docstring of the function IMO. Or it should be reworded to be more like a docstring.

"This function..."

**Comment by HarelM** [comment]
Another library seems like an overkill in terms of maintenance at this point, I'm not saying we shouldn't consider this in the future, but right now, in order to be able to move faster, keeping it here makes sense. 
Some of the methods are copied from other projects, but including all these references is a hassle as well, especially since we do want to work with "special" arrays from a performance perspective. 
I'll address the other comments. 
Thanks for the review!

File: ts/src/decoding/unpackNullableUtils.spec.ts (hunk 4 above)
Comment (HarelM) [reply ts/src/decoding/unpackNullableUtils.spec.ts:14]: Actually they are testing different things: they are testing that the reference is the same, i.e. that the original array was manipulated (zero copy) and that the values are the expected values.

File: ts/src/encoding/integerEncodingUtils.ts (hunk 3 above)
Comment (HarelM) [reply ts/src/encoding/integerEncodingUtils.ts:4]: No, so are RLE, zigzag, Delta and FastPFor, but we have their implementation here never the less, see my last comment on this thread.

File: mkdocs/mkdocs.yml [hunk 7]
```diff
@@ -50,7 +50,7 @@ markdown_extensions:
       hardbreak: True
       nbsp: True
   - admonition
-  - github-callouts
+  #- github-callouts
```
Comment (louwers) [inline mkdocs/mkdocs.yml]: If you want feel feel to create an issue to fix this and assign me.

Comment (HarelM) [reply mkdocs/mkdocs.yml]: I think this is resolved now, I need to test as the plugin author closed my bug...

Comment (HarelM) [reply mkdocs/mkdocs.yml]: I'll revert that, sorry...

File: ts/src/decoding/unpackNullableUtils.spec.ts (hunk 4 above)
Comment (louwers) [reply ts/src/decoding/unpackNullableUtils.spec.ts:14]: Makes sense!

File: ts/src/decoding/integerStreamDecoder.ts (hunk 6 above)
Comment (HarelM) [reply ts/src/decoding/integerStreamDecoder.ts:164]: This is an internal method of this file, but sure, I can improve the wording here...

File: mkdocs/mkdocs.yml (hunk 7 above)
Comment (HarelM) [reply mkdocs/mkdocs.yml]: reverted in 34f71b7

File: ts/src/decoding/integerStreamDecoder.ts (hunk 6 above)
Comment (HarelM) [reply ts/src/decoding/integerStreamDecoder.ts:164]: Fixed in 74e96a1

File: ts/src/decoding/unpackNullableUtils.spec.ts (hunk 5 above)
Comment (HarelM) [reply ts/src/decoding/unpackNullableUtils.spec.ts]: Resolved in 6fef2ed, I've also fixed some other places with a missing "the", there are probably others left if I needed to guess...

**Comment by HarelM** [comment]
Ok, one major step forward from coverage and test readability perspective.
